### PR TITLE
[WIP] Support to rust/wasm

### PR DESCRIPTION
**Pull request by albizures**
Hi, this PR is not complete yet, currently, I'm struggling with two things:

- the first one is when there is a change in the entry rust file, the new wasm file is being generated but the file in the build folder doesn't change. I think it's something related to the cache.
- and the second one is related to the secondary rust files when there is a change in these files parcel doesn't detect the change. I think I have to do something like in  StylusAssets, use a custom evaluator, but I'm not sure.

Any ideas?

Thanks in advance

related: #34

### CODE REVIEW

**Comment by devongovett** [comment]
Awesome start, really excited about this! I'll give some more feedback in a bit! 🎉

**Comment by softprops** [comment]
Awesome to find some one is _already_ working on this. So fast!

File: src/assets/RustAsset.js [hunk 1]
```diff
@@ -0,0 +1,93 @@
+const url = require('url');
+const path = require('path');
+const toml = require('toml');
+const {exec} = require('child-process-promise');
+
+const fs = require('../utils/fs');
+const JSAsset = require('./JSAsset');
+const localRequire = require('../utils/localRequire');
+
+const rustTarget = `wasm32-unknown-emscripten`;
```
Comment (devongovett) [inline src/assets/RustAsset.js]: I think maybe we'll want to use wasm32-unknown-unknown, which doesn't require emscripten. Maybe checkout [rustify](https://github.com/browserify/rustify) which does something similar for browserify, and [wasm-experiments](https://github.com/killercup/wasm-experiments) to see how it all works.

File: src/assets/RustAsset.js [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+    // Do nothing. Dependencies are collected by cargo :).
+  }
+
+  async parse(code) {
+    this.invalidateBundle();
+    this.invalidate();
+    this.dependencies.clear();
+    const {name, encoding, options} = this;
+    const release = process.env.NODE_ENV === 'production';
+    const cmd = `wargo build ${release ? ' --release' : ''}`;
```
Comment (devongovett) [inline src/assets/RustAsset.js]: did you mean `cargo`?

File: src/assets/RustAsset.js [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+const toml = require('toml');
+const {exec} = require('child-process-promise');
+
+const fs = require('../utils/fs');
+const JSAsset = require('./JSAsset');
+const localRequire = require('../utils/localRequire');
+
+const rustTarget = `wasm32-unknown-emscripten`;
+
+class RustAsset extends JSAsset {
```
Comment (devongovett) [inline src/assets/RustAsset.js]: I think we should probably make a `WASMAsset` class to load precompiled `.wasm` files, and use it as a base class for this one. That way the runtime can be shared across multiple wasm languages.

File: src/assets/RustAsset.js [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+  collectDependencies() {
+    // Do nothing. Dependencies are collected by cargo :).
+  }
+
+  async parse(code) {
+    this.invalidateBundle();
+    this.invalidate();
+    this.dependencies.clear();
```
Comment (devongovett) [inline src/assets/RustAsset.js]: why is this necessary?

File: src/assets/RustAsset.js [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+    );
+
+    const Module = {
+      // Path in the built project to the wasm file
+      wasmBinaryFile,
+      // Indicates that we are NOT running in node, despite 'require' being defined
+      ENVIRONMENT: 'WEB'
+    };
+
+    if (options.fromHtml) {
```
Comment (devongovett) [inline src/assets/RustAsset.js]: Why is it different when loaded from HTML?

File: src/assets/RustAsset.js (hunk 4 above)
Comment (albizures) [reply src/assets/RustAsset.js]: oh my bad, this was me trying to fix the rebuild as mentioned in the description, I'll remove it, thanks

File: src/assets/RustAsset.js (hunk 5 above)
Comment (albizures) [reply src/assets/RustAsset.js]: the difference is that with this approach you can use rust as a standalone executable, without javascript.

I got the idea from this [article](https://hoverbear.org/2017/04/06/the-path-to-rust-on-the-web/)

File: src/assets/RustAsset.js (hunk 2 above)
Comment (albizures) [reply src/assets/RustAsset.js]: no, wargo is a library that setup your environment and makes easy to compile rust into wasm

File: src/assets/RustAsset.js (hunk 1 above)
Comment (albizures) [reply src/assets/RustAsset.js]: right now I'm using wargo and for now, wargo just uses wasm32-unknown-emscripten 

but sure, I'm going to take a look at rustify

File: src/assets/RustAsset.js (hunk 2 above)
Comment (Phrohdoh) [reply src/assets/RustAsset.js]: Does this PR verify that wargo is installed and accessible or fail violently?

Comment (albizures) [reply src/assets/RustAsset.js]: right now it's failing violently, 😄  but I was working on an api more friendly to use wargo directly instead of using wargo through child_process. 
However, I was thinking about to stop using wargo since wargo is using `wasm32-unknown-emscripten` and I think `wasm32-unknown-unknown` is a better option...

**Comment by shawwn** [comment]
Hey, I can set aside some time over the holidays to help finish the PR if you'd like. If you want a hand, ping me in parcel's Slack.

**Comment by albizures** [comment]
Sure thing @shawwn thanks!

**Comment by albizures** [comment]
Hey @devongovett, I made some changes

So, I got rid of wargo, and now I'm using `wasm32-unknown-unknown` directly also these changes fix my first point in the description 🎉 

what is next? 
- tackle my second point in the description
- validate the existence of `rustc`, `wasm32-unknown-unknown` target and create a useful alert
- use [wasm-gc](https://github.com/alexcrichton/wasm-gc), in production mode, if it exists

Do you have something else in mind?

also, I made a [demo](https://github.com/albizures/parcel-rust)

cc: @shawwn

**Comment by devongovett** [comment]
🎉 Awesome work @albizures! Your list sounds good.

I think you can use `rustc` to determine the list of imported files: use the `--emit dep-info` flag and it will generate a file with all the deps in it that we could parse. These can be sent to parcel as dependencies with the `includedInParent: true` flag, which will cause parcel to watch those files but not try to parse them as they've already been compiled by rust.

About validating the existence of `rustc`, I wonder if we could just install it for you if you don't have it installed. Especially since we depend on the `nightly` toolchain, a specific backend (`wasm32-unknown-unknown`), and `wasm-gc`, we could just install those for you if they aren't already. We do this for other compilers like typescript, but it's slightly easier since those are in npm.

Also, we should probably support using `cargo build` rather than `rustc` directly, if we detect a `Cargo.toml` file. This will allow some more complex builds to happen. Not sure if we can detect dependencies in that case though.

Let's also start thinking about what we want the API for importing wasm modules to be. I think there should be a small runtime built into Parcel to load and instantiate wasm modules. Basically we should take care of the `WebAssembly.instantiate` call and the import should resolve to `instance.exports`.

There are two cases to think about.

Synchronous import, wasm is either inlined into JS or somehow loaded prior to the JS running:

```javascript
import {add} from './add.rs';

console.log(add(2, 3));
```

Asynchronous import, wasm is placed in a separate file and loaded on demand:

```javascript
let {add} = await import('./add.rs');

console.log(add(2, 3));
```

Sorry for the long comment. All of this definitely doesn't need to happen in this PR, just getting the ideas flowing! 😄

**Comment by devongovett** [comment]
Going to merge this into the `wasm` branch on the main repo. I've added you as a collaborator @albizures so you should be able to work off that branch.

**Comment by albizures** [comment]
oh great, thanks 🎉 

and I agree with you with all of your points.

I'll let you know any news
